text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,513 @@
+
+![Request-Response](images/otlp-request-response.png)
+
+_Note: this protocol is concerned with reliability of delivery between one pair of client/server nodes and aims to ensure that no data is lost in-transit between the client and the server. Many telemetry collection systems have multiple nodes that the data must travel across until reaching the final destination (e.g. application -> agent -> collector -> backend). End-to-end delivery guarantees in such systems is outside of the scope of OTLP. The acknowledgements described in this protocol happen between a single client/server pair and do not span multiple nodes in multi-hop delivery paths._
+
+### Client Modes
+
+OTLP defines 2 modes of operation for clients: Unary and Streaming.
+
+Unary mode is simpler to implement and provides good performance in sub-millisecond latency networks, at the cost of worse throughput in high latency networks.
+
+Streaming mode requires somewhat more complicated implementation but provides significantly better throughput on high latency networks. 
+
+Clients are recommended to implemented both modes but may choose to implement only one.
+
+Servers MUST implement both modes. Clients operating in different modes can connect to the same server.
+
+#### Unary
+
+In Unary mode after sending the request the client MUST wait until the response is received from the server. There can be at most only one request in flight that is not yet acknowledged by the server. The server MUST reply with `Export` response that includes the ID of the last request. Until such response or an error from underlying transport layer is received the client MUST NOT send any other requests. Unary mode is implemented using gRPC unary RPC.
+
+![Unary](images/otlp-unary.png)
+
+Unary mode is recommended when simplicity of implementation is desirable and when the client and the server are connected via very low-latency network, such as for example when the client is an instrumented application and the server is a OpenTelemetry Service running as a local daemon.
+
+The maximum achievable throughput in unary mode is `max_request_size / (network_latency + server_response_time)`. For example if the request can contain at most 100 spans, network roundtrip latency is 200ms and server response time is 300 ms, then the maximum achievable throughput is `100 spans / (200ms+300ms)` or 200 spans per second. It is easy to see that in high latency networks or when the server response time is high it is difficult to achieve good throughput unless requests are very big (even when client, server and network throughput potentially allow much bigger processing rate).
+
+The ID field of the Export request and response is not used in Unary mode since the underlying transport already supports acknowledgements and one-to-one mapping of requests and responses.
+
+#### Streaming
```

Comment:
I'd like to ask the "why streaming?" question in a slightly different way than rnburn did. What benefits are you getting from a streaming model that you wouldn't also get from a unary model that allows more than one in-flight request?

There are a variety of drawbacks to the pattern of "implement my own request/response pairing over a single streaming RPC." Some of them you've listed in the doc already.

* You have to implement your own bookkeeping on client and server to pair up requests and responses.

* Things like metrics and tracing at the level of the RPC framework become much less useful in a streaming model. For example, in a unary RPC model, I can look at the normal request counters and error counters and latency metrics and so forth to understand what's going on. When multiple logical RPCs are multiplexed over a single stream, that information is lost and the client and server have to re-implement all of those metrics by hand (with non-standard names, etc).

* I don't really trust loadbalancing of streaming RPCs. I know you ran some tests, and you rightfully talk in the doc about the need to periodically close streams, but my experience is that over time you end up turning that knob so far down that you're more or less approximating unary RPCs anyways.

I believe that GRPC (unlike the Google systems that it was based on) has connection-level flow control, which historically was the main reason to prefer streaming RPCs over unary RPCs. So it's not entirely clear to me what benefit is coming from the approach proposed here. Thoughts?

In my experience, the two motivations where streaming RPCs are useful are:

* You need RPC-level flow control (not just connection-level flow control).

* You need to pin some kind of session state on a particular server, i.e. you have several different request messages that *must* be sent to the same server process (and you don't want to do you own bookkeeping to make that happen).

It doesn't seem like either of those applies here though.

Reply:
>I'd like to ask the "why streaming?" question in a slightly different way than rnburn did. What benefits are you getting from a streaming model that you wouldn't also get from a unary model that allows more than one in-flight request?

@akehlenbeck thank you for thoughful comments.

I did consider doing Unary with concurrent requests and I actually have a simple implementation here https://github.com/tigrannajaryan/exp-otelproto/tree/master/protoimpls/grpc_unary_async (it is part of the benchmarks).

2 reasons why I chose streaming after comparing it with concurrent Unary:

1. Ability to optimize the Resource field and avoid sending it with every request if it doesn't change during streaming session. This optimization is not specified in this document, but I intend to add it. Unfortunately it is hard to quantify the benefits of the optimization without knowing the ratio and sizes of Resources vs Spans sent in a single session however this capability was part of OpenCensus protocol and I wanted to keep this optimization possibility. For this to be possible one need to be able to send a sequence of requests to the same server behind load balancer and streaming allows to do that.

2. Streaming consumes slightly less CPU than Unary per span sent (about 3% in benchmarks). This is small but nevertheless was visible in the benchmarks of particular implementation.

I want to note that my preference for streaming is not a strong opinion. I am open to re-considering and dropping streaming in favour of concurrent unary, especially if I hear voices telling that Resource optimization is unneeded.

>I don't really trust loadbalancing of streaming RPCs.

Can you expand on this? I am interested in knowing what were the problems that you observed.

>I believe that GRPC (unlike the Google systems that it was based on) has connection-level flow control, which historically was the main reason to prefer streaming RPCs over unary RPCs. So it's not entirely clear to me what benefit is coming from the approach proposed here. Thoughts?

Ability for the server to send acknowledgements out of order is one benefit with using own flow control rather than rely on stream's built-in flow control. Compared to unary obviously there is no flow control benefits, unary already allows out of order responses.

>In my experience, the two motivations where streaming RPCs are useful are:
>...
>You need to pin some kind of session state on a particular server, i.e. you have several different request messages that must be sent to the same server process (and you don't want to do you own bookkeeping to make that happen).

This one does apply in our case, that's how we would do Resource optimization.